kubectl-plugin/pkg/cmd/submit/submit.go

```diff
@@ -0,0 +1,357 @@
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Working dir path does not exist. Failed with: %w", err)
+		} else if err != nil {
+			return fmt.Errorf("Error occurred when checking working dir file: %w", err)
+		} else if !info.IsDir() {
+			return fmt.Errorf("Working dir is not a directory, please use directory: %w", err)
+		}
+
+		// RayJob yaml validation
+		info, err = os.Stat(options.fileName)
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Ray Job file does not exist. Failed with: %w", err)
+		} else if err != nil {
+			return fmt.Errorf("Error occurred when checking ray job file: %w", err)
+		} else if !info.Mode().IsRegular() {
+			return fmt.Errorf("Filename given is not a regular file. Failed with: %w", err)
+		}
+
+		err = decodeRayJobYaml(options.fileName, options.RayJob)
+		if err != nil {
+			return fmt.Errorf("Failed to decode RayJob Yaml: %w", err)
+		}
+
+		submissionMode := options.RayJob.Object["spec"].(map[string]interface{})["submissionMode"]
+		if submissionMode != nil {
+			if submissionMode == rayv1api.K8sJobMode || submissionMode == rayv1api.HTTPMode {
+				return fmt.Errorf("Submission mode of the Ray Job is not supported")
+			}
+		}
+	}
+
+	if len(options.runtimeEnv) > 0 {
+		info, err := os.Stat(options.runtimeEnv)
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Runtime Env file does not exist. Failed with: %w", err)
+		} else if err != nil {
+			return fmt.Errorf("Error occurred when checking runtime env file: %w", err)
+		} else if !info.Mode().IsRegular() {
+			return fmt.Errorf("Filename given is not a regular file. Failed with: %w", err)
+		}
+	}
+	return nil
+}
+
+func (options *SubmitJobOptions) Run(ctx context.Context, factory cmdutil.Factory) error {
+	k8sClients, err := client.NewClient(factory)
+	if err != nil {
+		return fmt.Errorf("failed to initialize clientset: %w", err)
+	}
+
+	if options.RayJob != nil {
+		// createdRayJob, err = k8sClients.CreateRayCustomResource(ctx, util.RayJob, options.configFlags.Namespace, unstructuredRayjob)
+		options.RayJob, err = k8sClients.DynamicClient().Resource(util.RayJobGVR).Namespace(*options.configFlags.Namespace).Create(ctx, options.RayJob, v1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("Error when creating RayJob CR: %w", err)
+		}
+	}
+
+	var resourceType util.ResourceType
+	var resourceName string
+	if len(options.cluster) > 0 {
+		resourceType = util.RayCluster
+		resourceName = options.cluster
+	} else if options.RayJob != nil && len(options.RayJob.GetName()) > 0 {
+		resourceType = util.RayJob
+		resourceName = options.RayJob.GetName()
+	} else {
+		return fmt.Errorf("Unknown cluster and did not provide Ray Job. One of the fields must be set")
+	}
+
+	svcName, err := k8sClients.GetRayHeadSvcName(ctx, *options.configFlags.Namespace, resourceType, resourceName)
+	if err != nil {
+		return fmt.Errorf("Failed to find service name: %w", err)
+	}
+
+	portForwardCmd := portforward.NewCmdPortForward(factory, *options.ioStreams)
```

Comment:
Should there be some polling logic to wait for the RayCluster to be ready before we attempt the next step to port-forward and submit a job?

Reply:
You're right, we should. Adding check for cluster status.

Reply:
Added a check for cluster status and keep waiting until it's ready. What would be the maximum amount of time we want to wait for? (currently set to 30 seconds)

Reply:
Changed to 2 minutes for more buffer time.